### ISSUE TITLE
[Bug]:Structures/Mobs that still struggle to spawn with modded biomes (Terralith/BYG)

### ISSUE THREAD

**Issue by DoctorN0va** [Alex-the-666/Ice_and_Fire#4920]
### What happened?

- Hydra Dens
- Stymphalian Birds
- Sirens(?)
- Trolls
- Cyclops(Both wandering and cave)
- Dragon caves (Known)

### Minecraft Version

1.18

### Ice and Fire version

2.1.12 beta 3

### Citadel version

1.11.3

### Other Information

- [X] I am using other mods besides Ice and Fire
- [X] I have attached a log file from the session that I have experienced this issue

**Comment by gummie2022** [comment]
Hippocampuses, Sea Serpents, Desert Myrmexes, Death Worms, and Pixies do not spawn in Terralith/Byg-Modified Vanilla biomes.

Hydra Lair, Jungle Myrmex, and Siren Island do not spawn in normal vanilla biomes which likely because of Caves & Cliffs biome overhaul.

**Comment by TheBv** [comment]
Should've all been addressed. I was a bit silly when integrating them and at some point copy pasted things incorrectly. (Used `BIOME_CATEGORY` instead of `BIOME_REGISTRY`)
I've tested it myself and each feature/structure now has at least one dedicated terralith biome. In general vanilla minecraft biomes still exist as well for which things generate as well.
I'll add a couple more biomes which I might've missed for BYG but besides that I'm pretty sure all these issues should've been addressed.
Due to the fact that there's more biomes though the generation features might feel a bit off. But there's sadly not much I can do about this. 
Also the cave issues still exist and will continue to exist until that system is reworked.
 
EDIT: I'll keep this open for now in case people have anything to add.

**Comment by gummie2022** [comment]
so you supposedly going to use `REGISTRY_NAME` instead of `BIOME_CATEGORY`?

**Comment by gummie2022** [comment]
also, `terralith:cave/ice_caves` does not exist and i have no idea why it's still on Ice Dragon Biome config (both Lightning Dragon and Ice Dragon Roost/Cavern)

and `lightning_dragon_cave_biomes` config is clearly accidentally copypasted on `ice_dragon_cave_biomes` config lmao

**Comment by TheBv** [comment]
> also, `terralith:cave/ice_caves` does not exist and i have no idea why it's still on Ice Dragon Biome config (both Lightning Dragon and Ice Dragon Roost/Cavern)

Yeah I just read up about that on a wiki apparently it was scrapped almost immediately after having been introduced. Which is said since I'd have really liked them to exist ^^

**Comment by gummie2022** [comment]
I suggest talking with the Terralith/Byg Dev-Team to sort these mob spawning problems out.

**Comment by gummie2022** [comment]
> Hippocampuses, Sea Serpents, Siren, Stymphalian Birds, Death Worms, Cyclops, and Pixies do not spawn in Terralith-Modified Vanilla biomes.

Most IAF mobs in Terralith biomes, including Terralith-modified Vanilla biomes, can't spawn even when I attempted to adjust the biome config. Funny enough, it works fine with BYG biomes, according to Gack in Discord which is unusual that few people suggest it should have paired with Terralith to help them spawn seamlessly.

Also, it's best to sort this up with Terralith's end before they update their mod to 2.2.5

**Comment by catter1** [comment]
Hi, I'm one of the devs that works on Terralith. As you can probably tell by the number of issues this repository, but I do see a lot of complaints about IaF structures and mobs not really generating much in Terralith biomes. What needs to be done to improve compatibility between our two mods? I'm happy to do some things on Terralith's end if it's needed.

As a side note, I've seen in these issues that some mobs and structures *do* spawn, but they are quite rare, or not all of them do. That also matches what I see in our Discord. Either way, something needs to be improved.

In the current 2.2.4 version, we started registering the biomes in the Forge biome registry, and made it compatibile with Terrablender. However - and I don't know how - I somehow forgot to add the biome tags back into it.

> Also, it's best to sort this up with Terralith's end before they update their mod to 2.2.5

So, I do plan on releasing a 2.2.5 version of Terralith with the tags added back, and potentially some other unrelated fixes. But before I do that, I would like to help improve compat with IaF if possible. So, @TheBv (or anyone else who is knowledgable for that matter), please let me know if there's anything I can do! I'm `@catter1` on Discord if you prefer that, or I will check on this issue.

**Comment by TheBv** [comment]
@catter1 
We don't really need anything else besides the forge tags. 
Basically per default a biome needs to meet certain tags for it to be added as a possible candidate for a structure/feature.
If you could you could also add entries to the `BIOME_DICT` (not sure how that works though) and they were removed in 1.19 onwards as well, it'd just be a bonus for the 1.18 version.

I guess the biggest challenge there is, besides the missing tags, is the fact that all of the caves (besides a few very small special ones) all have the same biome(default biomes as far as I can remember have normal biomes underneath but it's been a while since I last checked).
So we can't easily spawn ice dragon caves in biomes that are cold, because the cave biomes are all the same.
I personally don't see an easy way for you to fix this though. I think it would be worth investigating since this also means that emerald ore can't generate in these caves, or possibly other custom mod ore that is generated based on the biome (although I assume there aren't a lot). I might be wrong about this though.

Anyway long story short: Tags are all we need, making the default cave biomes the same as the surface biome would be a huge plus.
In the future Ice and Fire might solve that issue itself. I just don't have the time at all right now.

**Comment by catter1** [comment]
What do you mean by "all of the caves all have the same biome"? Each cave biome is its own, well, biome, and added to the biome distribution and registry the same as surface biomes. Each cave biome should have biome tags custom assigned for each as well. I can double check that, but for example, Frostfire Caves would be `is_cold` (or whatever the tag is called).

Otherwise, sounds good. When I said "adding biome to the Forge biome registry", I *believe* that also means adding them to `BIOME_DICT`. But since I am not much of a modder, forgive me if I am wrong. Anyways, I will add the tags, and do some testing myself to make sure everything is working. Thanks for your reply!

**Comment by TheBv** [comment]
Very late on the reply here. Somehow I missed this; sorry.
Not sure if I'm too late but here's what I meant:
Let's say we have a custom biome e.g `terralith:shrubland` for some reason the cave underneath won't have the biome `terralith:shrubland` like how it would be for default minecraft biomes. Instead it will have the biome `terralith:cave` (I think or something generic like that at least).
So while the shrubland might have certain tags the cave underneath won't.

Not sure if that's too late and maybe you already fixed/changed this I just though I might as well follow up on this.

**Comment by catter1** [comment]
Yeah, caves are separate, so will have their own tags. Treated like any other normal biome.

I was busy over the summer so never got around to releasing the fix for this, but I hope to quite soon. New version should have everything tagged as intended

**Comment by Sideralis001** [comment]
I have a clue of what could happen, in the mod configuration in config\iceandfire there are files with a .json ending, I changed some values that had these to true, I don't know why it happens but when these values are false it does not generate entities, and I don't know why by default they are like this, I hope they can solve this
  {
         "type": "REGISTRY_NAME",
         "negate": false,
         "value": "byg:autumnal_valley"
       }

I changed the negate variable from false to true, with this mobs began to appear in that biome

**Comment by SiverDX** [comment]
See https://github.com/AlexModGuy/Ice_and_Fire/issues/4934#issuecomment-1891763336

**Comment by Xplodin** [comment]
How would this work for datapack / custom dimensions?

Not twilight forest. but like ad astra

**Comment by tuw9804-ctrl** [comment]
@TheBv  is it fixed in the version Oh-The-Biomes-Weve-Gone-NeoForge-2.6.0
